extensions/standard-processors/processors/TailFile.cpp

```diff
@@ -403,102 +435,136 @@ void TailFile::onTrigger(const std::shared_ptr<core::ProcessContext> &context, c
-          std::string logName = baseName + "." + std::to_string(state.second.currentTailFilePosition_) + "-" + std::to_string(state.second.currentTailFilePosition_ + flowFile->getSize()) + "."
-              + extension;
-          flowFile->updateKeyedAttribute(FILENAME, logName);
-          state.second.currentTailFilePosition_ += flowFile->getSize();
-          storeState(context);
-        }
-      }
-      state.second.currentTailFileModificationTime_ = ((uint64_t) (statbuf.st_mtime) * 1000);
-    } else {
-      logger_->log_warn("Unable to stat file %s", fullPath);
+  if (!session->existsFlowFileInRelationship(Success)) {
+    yield();
+  }
+}
+
+void TailFile::processFile(const std::shared_ptr<core::ProcessContext> &context,
+                           const std::shared_ptr<core::ProcessSession> &session,
+                           const std::string &fileName,
+                           TailState &state) {
+  if (utils::file::FileUtils::file_size(state.fileNameWithPath()) < state.position_) {
+    processRotatedFiles(context, session, state);
+  }
+
+  processSingleFile(context, session, fileName, state);
+}
+
+void TailFile::processRotatedFiles(const std::shared_ptr<core::ProcessContext> &context,
+                                   const std::shared_ptr<core::ProcessSession> &session,
+                                   TailState &state) {
+    std::vector<TailState> rotated_file_states = findRotatedFiles(state);
+    for (TailState &file_state : rotated_file_states) {
+      processSingleFile(context, session, file_state.file_name_, file_state);
     }
+    state.position_ = 0;
+    state.checksum_ = 0;
+}
+
+void TailFile::processSingleFile(const std::shared_ptr<core::ProcessContext> &context,
+                                 const std::shared_ptr<core::ProcessSession> &session,
+                                 const std::string &fileName,
+                                 TailState &state) {
+  std::string full_file_name = state.fileNameWithPath();
+
+  if (utils::file::FileUtils::file_size(full_file_name) == 0u) {
+    logger_->log_warn("Unable to read file %s as it does not exist or has size zero", full_file_name);
+    return;
+  }
+  logger_->log_debug("Tailing file %s from %llu", full_file_name, state.position_);
+
+  std::size_t last_dot_position = fileName.find_last_of('.');
+  std::string baseName = fileName.substr(0, last_dot_position);
+  std::string extension = fileName.substr(last_dot_position + 1);
+
+  if (!delimiter_.empty()) {
+    char delim = delimiter_[0];
+    logger_->log_trace("Looking for delimiter 0x%X", delim);
+
+    std::vector<std::shared_ptr<FlowFileRecord>> flowFiles;
+    uint64_t checksum = state.checksum_;
+    session->import(full_file_name, flowFiles, state.position_, delim, true, &checksum);
+    logger_->log_info("%u flowfiles were received from TailFile input", flowFiles.size());
+
+    for (auto &ffr : flowFiles) {
```

Comment:
I don't know -- if I was reading this code and it had `const auto &`, my mind would go "wait, what, how is this const? we will modify the flow file... ah OK, it's a const pointer to non-const".  Leaving it as `auto &` was an attempt to spare others this mental detour, but I understand it can create other mental detours. :)

I will change it if you want me to, just checking if you really do.

Reply:
The difference is clear to me, but I can only speak for myself. I'm OK with leaving it as is.

Reply:
I would agree with @fgerlits, I don't know how often we actually modify an `std::shared_ptr<FlowFile>`, but if it is negligible I would go as far to replace all these with a `FlowFilePtr` with more sensible (for me) pointer access logic, i.e. `const T& operator->() const` (and we can also add a `FlowFileConstPtr` when needed)